federation/pkg/federation-controller/service/dns.go

```diff
@@ -254,6 +254,9 @@ func (s *ServiceController) ensureDnsRecords(clusterName string, cachedService *
 	// the state of the service when we last successfully sync'd it's DNS records.
 	// So this time around we only need to patch that (add new records, remove deleted records, and update changed records.
 	//
+	if s.dns == nil {
+		return nil
```

Comment:
I think that your code should not call ensureDnsRecords if the DNS provider is not initialized.  And if it does, ensureDnsRecords should return an error (as it originally did), not nil like this.  But again, I can fix this in a followup PR if you prefer.